File: android/src/main/java/name/zeno/android/presenter/input/InputActivity.java
package name.zeno.android.presenter.input;

import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.ColorInt;
import android.support.design.widget.TextInputEditText;
import android.support.v7.widget.AppCompatButton;
import android.text.InputType;
import android.view.Gravity;

import java.util.regex.Pattern;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnTextChanged;
import name.zeno.android.data.models.TextData;
import name.zeno.android.presenter.Extra;
import name.zeno.android.presenter.ZActivity;
import name.zeno.android.system.ZStatusBar;
import name.zeno.android.util.R;
import name.zeno.android.util.R2;
import name.zeno.android.util.ZDimen;
import name.zeno.android.util.ZRex;
import name.zeno.android.util.ZString;
import name.zeno.android.widget.SimpleActionbar;

/**
 * @author 陈治谋 (513500085@qq.com)
 * @since 2017/8/14
 */
public class InputActivity extends ZActivity
{

  @BindView(R2.id.et_content) TextInputEditText etContent;
  @BindView(R2.id.actionbar)  SimpleActionbar   actionbar;
  @BindView(R2.id.btn_submit) AppCompatButton   submitBtn;

  private Pattern  pattern;
  private TextData textData;

  @Override
  protected void onCreate(Bundle savedInstanceState)
  {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_input);
    ButterKnife.bind(this);

    textData = Extra.getData(getIntent());

    if (textData.isWhiteStatusBar()) {
      ZStatusBar.INSTANCE.lightMode(this);
      actionbar.setBackgroundColor(Color.WHITE);
      actionbar.setStatusBarViewBackgroundColor(Color.WHITE);
      @ColorInt int color = Color.parseColor("#333333");
      actionbar.setTitleColor(color);
      actionbar.setPreTint(color);
    }

    if (textData.getBtnBackground() != 0) {
      submitBtn.setBackgroundResource(textData.getBtnBackground());
    }

    textData.setResult("");
    if (ZString.notEmpty(textData.getRegex())) {
      pattern = Pattern.compile(textData.getRegex());
    }

    actionbar.setTitleText(textData.getTitle());

    switch (textData.getType()) {
      case TextData.TYPE_TEXT:
        break;
      case TextData.TYPE_PHONE:
      case TextData.TYPE_NUMBER:
        etContent.setInputType(InputType.TYPE_CLASS_NUMBER);
        break;
      case TextData.TYPE_CONTENT:
        etContent.setSingleLine(false);
        etContent.setHeight(ZDimen.dp2px(120));
        etContent.setGravity(Gravity.NO_GRAVITY);
        break;
    }
    etContent.setHint(textData.getHint());
    etContent.setText(textData.getPreFill());
    etContent.setSelection(etContent.getText().length());
    etContent.setOnEditorActionListener((textView, i, keyEvent) -> {
      onClickSubmit();
      return true;
    });
  }

  @OnTextChanged(R2.id.et_content) void onContentChange(CharSequence s)
  {
    textData.setResult(s.toString());
  }

  @OnClick(R2.id.btn_submit) void onClickSubmit()
  {
    if (textData.getType() == TextData.TYPE_PHONE && !ZRex.validPhone(textData.getResult())) {
      snack("手机号码格式不正确");
      return;
    }

    if (pattern != null && !pattern.matcher(textData.getResult()).matches()) {
      snack(textData.getRegexHint());
      return;
    }

    setResult(RESULT_OK, Extra.setData(textData));
    finish();
  }

}

File: library/src/main/java/name/zeno/android/presenter/input/InputActivity.kt
package name.zeno.android.presenter.input

import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.text.InputType
import android.view.Gravity
import kotlinx.android.synthetic.main.activity_input.*
import name.zeno.android.data.models.TextData
import name.zeno.android.presenter.ZActivity
import name.zeno.android.system.ZStatusBar
import name.zeno.android.util.R
import name.zeno.android.util.ZDimen
import name.zeno.android.util.ZEditor
import name.zeno.android.util.ZRex
import name.zeno.android.widget.ZTextWatcher
import java.util.regex.Pattern

/**
 * @param data [TextData]
 * @return data [TextData]
 *
 * @author 陈治谋 (513500085@qq.com)
 * @since 2017/8/14
 */
class InputActivity : ZActivity() {

  private var pattern: Pattern? = null
  private lateinit var textData: TextData

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_input)

    ZTextWatcher.watch(et_content, { _, s -> textData.result = s })
    btn_submit.setOnClickListener { submit() }

    textData = intent.getParcelableExtra("data")

    if (textData.isWhiteStatusBar) {
      ZStatusBar.lightMode(this)
      actionbar.setBackgroundColor(Color.WHITE)
      actionbar.setStatusBarViewBackgroundColor(Color.WHITE)
      val color = Color.parseColor("#333333")
      actionbar.setTitleColor(color)
      actionbar.setPreTint(color)
    }

    if (textData.btnBackground != 0) {
      btn_submit.setBackgroundResource(textData.btnBackground)
    }

    textData.result = ""
    if (textData.regex.isNullOrEmpty().not()) {
      pattern = Pattern.compile(textData.regex)
    }

    actionbar.setTitleText(textData.title)

    when (textData.type) {
      TextData.TYPE_TEXT -> {
      }
      TextData.TYPE_PHONE, TextData.TYPE_NUMBER -> et_content.inputType = InputType.TYPE_CLASS_NUMBER
      TextData.TYPE_CONTENT -> {
        et_content.setSingleLine(false)
        et_content.height = ZDimen.dp2px(120f)
        et_content.gravity = Gravity.NO_GRAVITY
      }
    }
    et_content.hint = textData.hint
    et_content.setText(textData.preFill)
    et_content.setSelection(et_content.text?.length ?: 0)
    ZEditor.actionDone(et_content) { submit() }
  }


  private fun submit() {
    if (textData.type == TextData.TYPE_PHONE && !ZRex.validPhone(textData.result)) {
      snack("手机号码格式不正确")
      return
    }

    if (pattern != null && !pattern!!.matcher(textData.result).matches()) {
      snack(textData.regexHint ?: "")
      return
    }

    setResult(RESULT_OK, Intent().putExtra("data", textData))
    finish()
  }
}
